include/picongpu/fields/incidentField/profiles/InsightPulse.hpp

```diff
@@ -0,0 +1,618 @@
+                         */
+                        HINLINE InsightPulseFunctorIncidentE(float_X const currentStep, float3_64 const unitField)
+                            : timeOriginPIC(currentStep * DELTA_T)
+                        {
+                            // load data at timestep 0
+                            auto& openPMDdata = OpenPMDdata<T_Params>::get();
+
+                            // get field data
+                            fieldDataBox = openPMDdata.bufferFieldData->getDeviceBuffer().getDataBox();
+
+                            // get field data attributes
+                            extentOpenPMDdataBox = openPMDdata.bufferExtentOpenPMD->getDeviceBuffer().getDataBox();
+                            cellSizeOpenPMDdataBox = openPMDdata.bufferCellSizeOpenPMD->getDeviceBuffer().getDataBox();
+                            offsetOpenPMDdataBox = openPMDdata.bufferOffsetOpenPMD->getDeviceBuffer().getDataBox();
+                        }
+
+                        /** Read incident field E value for the given position and time step
+                         *
+                         * @param totalCellIdx cell index in the total domain (including all moving window slides)
+                         * @return incident field E value in internal units
+                         */
+                        HDINLINE float3_X operator()(floatD_X const& totalCellIdx) const
+                        {
+                            return getPolarisationVector() * getValueE(totalCellIdx);
+                        }
+
+                        //! Get a unit vector with linear E polarization
+                        HDINLINE static constexpr float3_X getPolarisationVector()
+                        {
+                            return float3_X(Unitless::POL_DIR_X, Unitless::POL_DIR_Y, Unitless::POL_DIR_Z);
+                        }
+
+                        //! Get a 3-dimensional unit direction vector
+                        HDINLINE static constexpr float3_X getDirection()
+                        {
+                            return float3_X(Unitless::DIR_X, Unitless::DIR_Y, Unitless::DIR_Z);
+                        }
+
+                    private:
+                        /** Get value of E field for the given position
+                         * Linear interpolation of measured field data, which is aligned centered at the chosen
+                         * incident field plane. If the simulation window extent is greater than the field data chunk,
+                         * zero will be returned. In the other case, the field data chunk will be (transversally)
+                         * cropped.
+                         *
+                         * @param totalCellIdx cell index in the total domain (including all moving window slides)
+                         */
+                        HDINLINE float_X getValueE(floatD_X const& totalCellIdx) const
```

Comment:
This routine is very long. Can you split it a little in subroutines, e.g. outsourcing the index finding and interpolation?

Reply:
Done ✅